mlflow/pyfunc/backend.py

```diff
@@ -64,11 +65,22 @@ def serve(self, model_uri, port, host):
         # NB: Absolute windows paths do not work with mlflow apis, use file uri to ensure
         # platform compatibility.
         local_uri = path_to_local_file_uri(local_path)
-        command = ("gunicorn --timeout 60 -b {host}:{port} -w {nworkers} "
+
+        # Extract gunicorn_cmd_args and append
+        gunicorn_cmd_args = os.environ.get("GUNICORN_CMD_ARGS", '')
```

Comment:
@LarsDu I don't think all this work is necessary...the changes made by https://github.com/LarsDu/mlflow/pull/1/ should be sufficient. Having two spaces instead of one doesn't seem to affect the execution of the `gunicorn` entrypoint command....

Reply:
@dbczumar I have now tried out LarsDu#1. It does not work without some modification. I now need to go back to my day job, as I have already devoted several hours to this. Furthermore, I cannot run my test model (which takes longer than 60 seconds to load) on this open source repo without significant issues at this point, so someone will need to make a dummy model that takes a long time to load.  Please try the these changes yourself. I am still not sure about the consequences of passing two --timeout flags to gunicorn. I think in the end the better solution is to have an explicit `--timeout` flag as in my original PR.

Reply:
@LarsDu Can you provide information about the errors you ran into? I tested this locally by building a custom pyfunc model that sleeps for 80 seconds during the load process. I then tried serving the model with and without setting `GUNICORN_CMD_ARGS`.

Model code:
```
import mlflow
import mlflow.pyfunc

class Mod(mlflow.pyfunc.PythonModel):

    def __init__(self, delay):
        self.delay = delay

    def load_context(self, context):
        import time
        time.sleep(self.delay)
        print("Loaded context after: {} seconds".format(self.delay))

    def predict(self, ctx, model_input):
        return 7

mlflow.pyfunc.save_model(
    path="testmodel",
    python_model=Mod(delay=80)
)
```

Without setting `GUNICORN_CMD_ARGS`, the gunicorn worker timed out after 60 seconds, as expected:
```
$ mlflow models serve -m testmodel
2019/07/29 14:23:14 INFO mlflow.models.cli: Selected backend for flavor 'python_function'
2019/07/29 14:23:17 INFO mlflow.pyfunc.backend: === Running command 'source activate mlflow-baf9eadca7de081babb57c2fca984661a70e3461 1>&2 && gunicorn --timeout 60 -b 127.0.0.1:5000 -w 1 ${GUNICORN_CMD_ARGS} mlflow.pyfunc.scoring_server.wsgi:app'
[2019-07-29 14:23:17 -0700] [82493] [INFO] Starting gunicorn 19.9.0
[2019-07-29 14:23:17 -0700] [82493] [INFO] Listening at: http://127.0.0.1:5000 (82493)
[2019-07-29 14:23:17 -0700] [82493] [INFO] Using worker: sync
[2019-07-29 14:23:17 -0700] [82496] [INFO] Booting worker with pid: 82496
/Users/czumar/anaconda2/envs/mlflow-baf9eadca7de081babb57c2fca984661a70e3461/lib/python2.7/site-packages/mlflow/pyfunc/__init__.py:281: DeprecationWarning: .. Warning:: ``mlflow.pyfunc.load_pyfunc`` is deprecated since 1.0. This method will be removed in a near future release. Use ``mlflow.pyfunc.load_model`` instead.
  return load_pyfunc(model_uri, suppress_warnings)
[2019-07-29 14:24:18 -0700] [82493] [CRITICAL] WORKER TIMEOUT (pid:82496)
[2019-07-29 14:24:18 -0700] [82496] [INFO] Worker exiting (pid: 82496)
```

After passing `GUNICORN_CMD_ARGS="--timeout 85"`, the model loaded successfully without a worker timeout:
```
GUNICORN_CMD_ARGS="--timeout 85" mlflow models serve -m testmodel
2019/07/29 14:25:44 INFO mlflow.models.cli: Selected backend for flavor 'python_function'
2019/07/29 14:25:46 INFO mlflow.pyfunc.backend: === Running command 'source activate mlflow-baf9eadca7de081babb57c2fca984661a70e3461 1>&2 && gunicorn --timeout 60 -b 127.0.0.1:5000 -w 1 ${GUNICORN_CMD_ARGS} mlflow.pyfunc.scoring_server.wsgi:app'
[2019-07-29 14:25:47 -0700] [82587] [INFO] Starting gunicorn 19.9.0
[2019-07-29 14:25:47 -0700] [82587] [INFO] Listening at: http://127.0.0.1:5000 (82587)
[2019-07-29 14:25:47 -0700] [82587] [INFO] Using worker: sync
[2019-07-29 14:25:47 -0700] [82590] [INFO] Booting worker with pid: 82590
/Users/czumar/anaconda2/envs/mlflow-baf9eadca7de081babb57c2fca984661a70e3461/lib/python2.7/site-packages/mlflow/pyfunc/__init__.py:281: DeprecationWarning: .. Warning:: ``mlflow.pyfunc.load_pyfunc`` is deprecated since 1.0. This method will be removed in a near future release. Use ``mlflow.pyfunc.load_model`` instead.
  return load_pyfunc(model_uri, suppress_warnings)
Loaded context after: 80 seconds
```

Reply:
interesting. One second

Reply:
I am still getting:

```2019/07/29 14:49:14 INFO mlflow.models.cli: Selected backend for flavor 'python_function'
2019/07/29 14:49:14 INFO mlflow.pyfunc.backend: === Running command 'gunicorn --timeout 60 -b 127.0.0.1:5000 -w 1 ${GUNICORN_CMD_ARGS} mlflow.pyfunc.scoring_server.wsgi:app'
[2019-07-29 14:49:14 -0700] [55641] [INFO] Starting gunicorn 19.9.0
[2019-07-29 14:49:14 -0700] [55641] [INFO] Listening at: http://127.0.0.1:5000 (55641)
[2019-07-29 14:49:14 -0700] [55641] [INFO] Using worker: sync
[2019-07-29 14:49:14 -0700] [55644] [INFO] Booting worker with pid: 55644
[2019-07-29 14:49:14 -0700] [55644] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/arbiter.py", line 583, in spawn_worker
    worker.init_process()
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/workers/base.py", line 129, in init_process
    self.load_wsgi()
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/workers/base.py", line 138, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 52, in load
    return self.load_wsgiapp()
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/app/wsgiapp.py", line 41, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/Users/larryd/larryd_osx_venv/lib/python3.6/site-packages/gunicorn/util.py", line 350, in import_app
    __import__(module)
ModuleNotFoundError: No module named '${GUNICORN_CMD_ARGS}'
[2019-07-29 14:49:14 -0700] [55644] [INFO] Worker exiting (pid: 55644)
[2019-07-29 14:49:14 -0700] [55641] [INFO] Shutting down: Master
[2019-07-29 14:49:14 -0700] [55641] [INFO] Reason: Worker failed to boot.```

Reply:
I suspect this is due to the differences in the shells we are using. I am running these commands from OSX, which I believe has a different shell than what you may be using?

I believe different shells may not have the same support for the command `${{GUNICORN_CMD_ARGS}}`

Reply:
Ah, definitely interesting. I'm using a bash shell on OSX Mojave with gunicorn version `19.8.1`. Are you using a different shell?

Reply:
Ok it turns out our shells are exactly the same. It seems the fault is on my end. I erroneously added a whitespace after ${{GUNICORN_CMD_ARGS}} while removing the leading whitespace in the subsequent line while merging your PR (since had simultaneously written my own code). This causes the resulting error message above! I will revert everything back to your original PR, and then everything should be fine

Reply:
Awesome! Glad we got that figured out!

Reply:
Correction: There is supposed to be trailing and leading whitespace. Sigh